### PR TITLE
Fix/m2 3718 performance optimization

### PR DESCRIPTION
**Pull request by BamMironov**

### CODE REVIEW

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    appletDetails: AppletDetails,
+    completedEntity: CompletedEntity,
+  ) {
+    const inProgressEntities = this.getProgressState();
+    const localEndAt =
+      inProgressEntities?.[completedEntity.entityId]?.[completedEntity.eventId]
+        .endAt;
+    const serverEndAt = completedEntity.endAt;
+
+    const entityEventMissing =
```
[the remark this answers is not in the document]
Comment (anq83) [reply src/entities/applet/model/services/ProgressSyncService.ts:55]: What is missing and where? Locally or on server?

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      inProgressEntities?.[completedEntity.entityId]?.[completedEntity.eventId];
+    const isFlow = appletDetails.activityFlows.find(
+      flow => flow.id === completedEntity.entityId,
+    );
+
+    if (entityEventMissing) {
+      this.dispatch(
+        actions.completedEntityMissing({
```
Comment (anq83) [inline src/entities/applet/model/services/ProgressSyncService.ts:65]: Did you mean onCompletedEntityMissing ?

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+          eventId: completedEntity.eventId,
+        }),
+      );
+
+      this.logger.log(
+        `[RefreshService.refresh]: Added missing entity event [entityId: ${completedEntity.entityId}, eventId: ${completedEntity.eventId}]`,
+      );
+    }
+
+    const completedLaterThanFromServer = localEndAt && localEndAt < serverEndAt;
```
Comment (anq83) [inline src/entities/applet/model/services/ProgressSyncService.ts:82]: completedLaterThanOnServer

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          startAt: 0,
+          endAt: serverEndAt,
+          appletId: appletDetails.id,
+          entityId: completedEntity.entityId,
+          eventId: completedEntity.eventId,
+        }),
+      );
+
+      this.logger.log(
+        `[RefreshService.refresh]: Added missing entity event [entityId: ${completedEntity.entityId}, eventId: ${completedEntity.eventId}]`,
```
Comment (anq83) [inline src/entities/applet/model/services/ProgressSyncService.ts:78]: [RefreshService.syncAppletEntityWithServer] ...

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/entities/applet/model/services/ProgressSyncService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:55]: If the system can't find an entity-event by `entityId` and `eventId` in the global state then we consider it missing.

File: src/entities/applet/model/services/ProgressSyncService.ts (hunk 2 above)
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:65]: No :) `completedEntityMissing` is the correct name. Please check out [the official documentation](https://redux.js.org/style-guide/#model-actions-as-events-not-setters) with the Detailed Explanation section.

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  private dispatch: AppDispatch;
+  private state: RootState;
+
+  constructor(state: RootState, dispatch: AppDispatch, logger: ILogger) {
+    this.logger = logger;
+    this.dispatch = dispatch;
+    this.state = state;
+  }
+
+  private getProgressState() {
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:21]: ok

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      );
+
+      this.logger.log(
+        `[RefreshService.refresh]: Updated entity event [entityId: ${completedEntity.entityId}, eventId: ${completedEntity.eventId}] with data from the server`,
+      );
+    }
+  }
+
+  public sync(appletDto: AppletDetailsDto) {
+    return this.syncWithAppletDto(appletDto);
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:101]: yes

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+  private getProgressState() {
+    return selectInProgressEntities(this.state);
+  }
+
+  private async syncWithAppletDto(appletDto: AppletDetailsDto) {
+    const appletDetails = mapAppletDetailsFromDto(appletDto);
+    const fromDate = getMonthAgoDate();
+
+    const response = await EventsService.getCompletedEntities({
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:29]: ok

File: src/entities/applet/model/services/ProgressSyncService.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    );
+  }
+
+  private syncAppletEntityWithServer(
+    appletDetails: AppletDetails,
+    completedEntity: CompletedEntity,
+  ) {
+    const inProgressEntities = this.getProgressState();
+    const localEndAt =
+      inProgressEntities?.[completedEntity.entityId]?.[completedEntity.eventId]
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:51]: ok

File: src/entities/applet/model/services/ProgressSyncService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:55]: If the system cannot find an entity-event object on the global state by `entityId` and `eventId` then we consider it missing

File: src/entities/applet/model/services/ProgressSyncService.ts (hunk 4 above)
Comment (BamMironov) [reply src/entities/applet/model/services/ProgressSyncService.ts:78]: ok

**Review by anq83 — APPROVED** [review]

**Pull request merged**